pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,262 @@
+		return nil, fmt.Errorf("failed to get token for %s/%s : %v", t.Namespace, name, outErr)
+	}
+	return ret, nil
+}
+
+func (t *DynamicControllerClientBuilder) authenticate(saName, token string) (bool, error) {
+	// Try token review first
+	saNamespace := t.Namespace
+	username := apiserverserviceaccount.MakeUsername(saNamespace, saName)
+
+	tokenReview := &v1authenticationapi.TokenReview{Spec: v1authenticationapi.TokenReviewSpec{Token: token}}
+	if tokenResult, err := t.AuthenticationClient.TokenReviews().Create(tokenReview); err == nil {
+		if !tokenResult.Status.Authenticated {
+			klog.Warningf("Token for %s/%s did not authenticate correctly", saNamespace, saName)
+			return false, nil
+		}
+		if tokenResult.Status.User.Username != username {
+			klog.Warningf("Token for %s/%s authenticated as unexpected username: %s", saNamespace, saName, tokenResult.Status.User.Username)
+			return false, nil
+		}
+		klog.V(4).Infof("Verified credential for %s/%s", saNamespace, saName)
+		return true, nil
+	}
+
+	// If we couldn't run the token review, the API might be disabled or we might not have permission.
+	// Try to make a request to /apis with the token. If we get a 401 we should consider the token invalid.
+	clientConfigCopy := constructClient(saNamespace, saName, t.ClientConfig)
+	clientConfigCopy.BearerToken = token
+	clientConfigCopy.NegotiatedSerializer = legacyscheme.Codecs
+	client, err := restclient.UnversionedRESTClientFor(&clientConfigCopy)
+	if err != nil {
+		return false, err
+	}
+	err = client.Get().AbsPath("/apis").Do().Error()
+	if apierrors.IsUnauthorized(err) {
+		klog.Warningf("Token for %s/%s did not authenticate correctly: %v", saNamespace, saName, err)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (t *DynamicControllerClientBuilder) getOrCreateServiceAccount(name string) (*v1.ServiceAccount, error) {
+	sa, err := t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	if err == nil {
+		return sa, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Create the namespace if we can't verify it exists.
+	// Tolerate errors, since we don't know whether this component has namespace creation permissions.
+	if _, err := t.CoreClient.Namespaces().Get(t.Namespace, metav1.GetOptions{}); err != nil {
+		t.CoreClient.Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: t.Namespace}})
+	}
+
+	// Create the service account
+	sa, err = t.CoreClient.ServiceAccounts(t.Namespace).Create(&v1.ServiceAccount{ObjectMeta: metav1.ObjectMeta{Namespace: t.Namespace, Name: name}})
+	if apierrors.IsAlreadyExists(err) {
+		// If we're racing to init and someone else already created it, re-fetch
+		return t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	}
+	return sa, err
+}
+
+func getInt64Point(v int64) *int64 {
+	return &v
+}
+
+func (t *DynamicControllerClientBuilder) bearerTokenFunc(saName string) func() (string, error) {
+	return func() (string, error) {
```

Comment:
Can you somehow use the oauth2.TokenSource here?

https://github.com/kubernetes/kubernetes/blob/3e9b65b3cb9442b32e38a53f00619a8d5209e212/staging/src/k8s.io/client-go/transport/token_source.go#L33-L43